### PR TITLE
perf(compiler): skip type check and emit in bazel in some cases.

### PR DESCRIPTION
**Pull request by tbosch**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: packages/bazel/src/ngc-wrapped/emit_cache.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ng from '@angular/compiler-cli';
+import {fixUmdModuleDeclarations} from '@bazel/typescript';
+import * as tsickle from 'tsickle';
+import * as ts from 'typescript';
+
+interface EmitCachEntry {
```
Comment (gkalpak) [inline packages/bazel/src/ngc-wrapped/emit_cache.ts]: _EmitCachEntry_ --> _EmitCacheEntry_

File: packages/bazel/src/ngc-wrapped/emit_cache.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    const mergedEmitResult = tsickle.mergeEmitResults(targetSourceFiles.map(targetSourceFile => {
+      const targetGeneratedFile = program.getGeneratedFile(targetSourceFile.fileName);
+      const cacheSf = targetGeneratedFile ?
+          program.getTsProgram().getSourceFile(targetGeneratedFile.srcFileName) :
+          targetSourceFile;
+      const cacheEntry = getCache(cacheSf, targetGeneratedFile && targetGeneratedFile.genFileName);
+      if (cacheEntry) {
+        let useEmitCache = false;
+        if (targetGeneratedFile && !program.hasChanged(targetSourceFile.fileName)) {
+          // we emitted a GeneratedFile with the same content before -> use the cache
```
Comment (chuckjaz) [inline packages/bazel/src/ngc-wrapped/emit_cache.ts]: nit: /same content before/same content as before/

File: packages/compiler-cli/src/transformers/api.ts [hunk 3]
```diff
@@ -207,18 +207,33 @@ export interface CustomTransformers {
...
   cancellationToken?: ts.CancellationToken;
   emitOnlyDtsFiles?: boolean;
   customTransformers?: ts.CustomTransformers;
 }
 
-export interface TsEmitCallback { (args: TsEmitArguments): ts.EmitResult; }
+export interface EmitCallback { (args: EmitArguments): ts.EmitResult; }
+
+/**
+ * Represents a not yet emitted generated file.
```
Comment (chuckjaz) [inline packages/compiler-cli/src/transformers/api.ts]: Consider "Represents a generated file that has not yet been emitted".

File: packages/compiler-cli/src/transformers/api.ts [hunk 4]
```diff
@@ -207,18 +207,33 @@ export interface CustomTransformers {
...
+ */
+export interface GeneratedFile {
+  /**
+   * The file name of the sourceFile from which this file was generated.
+   */
+  srcFileName: string;
+  /**
+   * The file name of the generated but not yet emitted file.
+   * I.e. this ends with `.ts`.
```
Comment (chuckjaz) [inline packages/compiler-cli/src/transformers/api.ts]: I don't understand what this i.e. is an example of?

File: packages/compiler-cli/src/transformers/entry_points.ts [hunk 5]
```diff
@@ -12,3 +12,4 @@ import {CompilerHost, CompilerOptions, Program} from './api';
 
 export {createCompilerHost} from './compiler_host';
 export {createProgram} from './program';
+export {isGeneratedFile} from './util';
```
Comment (chuckjaz) [inline packages/compiler-cli/src/transformers/entry_points.ts]: Consider restoring the \n at the end of the file.

File: packages/compiler-cli/src/transformers/util.ts [hunk 6]
```diff
@@ -29,3 +30,7 @@ export function createMessageDiagnostic(messageText: string): ts.Diagnostic&Diag
     source: SOURCE,
   };
 }
+
+export function isGeneratedFile(fileName: string): boolean {
+  return GENERATED_FILES.test(fileName);
+}
```
Comment (chuckjaz) [inline packages/compiler-cli/src/transformers/util.ts]: Consider restoring \n at the end of the file.

**Review by chuckjaz — APPROVED** [review]
Some nits to consider.

File: packages/bazel/src/ngc-wrapped/emit_cache.ts (hunk 2 above)
Comment (tbosch) [reply packages/bazel/src/ngc-wrapped/emit_cache.ts]: Done

File: packages/compiler-cli/src/transformers/api.ts (hunk 3 above)
Comment (tbosch) [reply packages/compiler-cli/src/transformers/api.ts]: Done

File: packages/compiler-cli/src/transformers/api.ts (hunk 4 above)
Comment (tbosch) [reply packages/compiler-cli/src/transformers/api.ts]: Removed this.

File: packages/compiler-cli/src/transformers/entry_points.ts (hunk 5 above)
Comment (tbosch) [reply packages/compiler-cli/src/transformers/entry_points.ts]: Done

File: packages/compiler-cli/src/transformers/util.ts (hunk 6 above)
Comment (tbosch) [reply packages/compiler-cli/src/transformers/util.ts]: Done

**Pull request closed without merging**